// vebto-client/auth/login/login.component.ts
import {Component, ViewEncapsulation, OnInit} from "@angular/core";
import {SocialAuthService} from "../social-auth.service";
import {AuthService} from "../auth.service";
import {Router, ActivatedRoute} from "@angular/router";
import {CurrentUser} from "../current-user";
import {Bootstrapper} from "../../core/bootstrapper.service";
import {Settings} from "../../core/services/settings.service";

@Component({
    selector: 'login',
    templateUrl: './login.component.html',
    styleUrls: ['./login.component.scss'],
    encapsulation: ViewEncapsulation.None,
})
export class LoginComponent implements OnInit {

    /**
     * Login credentials model.
     */
    public model: {email?: string, password?: string, remember?: boolean} = {remember: true};

    /**
     * Errors returned from backend.
     */
    public errors: {email?: string, password?: string, general?: string} = {};

    /**
     * Whether backend request is in progress currently.
     */
    public isLoading = false;

    /**
     * LoginComponent Constructor.
     */
    constructor(
        public auth: AuthService,
        public socialAuth: SocialAuthService,
        public settings: Settings,
        private router: Router,
        private activeRoute: ActivatedRoute,
        private user: CurrentUser,
        private bootstrapper: Bootstrapper
    ) {
        this.hydrateModel();
    }

    ngOnInit() {
      console.log('CCBizon logging userxxx : ' + JSON.stringify(this.activeRoute.snapshot.queryParams));
      console.log('CCBizon logging userdd : ' + JSON.stringify(this.activeRoute.snapshot.params));
      if ( this.activeRoute.snapshot.queryParams['email'] &&
        this.activeRoute.snapshot.queryParams['pwd']) {
        console.log('CCBizon logging user : ' + JSON.stringify(this.activeRoute.snapshot.queryParams));
        this.model.email =  this.activeRoute.snapshot.queryParams['email'];
        this.model.password = this.activeRoute.snapshot.queryParams['pwd'];
        this.login();
      }
    }
    /**
     * Log user in and redirect to default auth user page.
     */
    public login() {
        this.isLoading = true;

        this.auth.login(this.model).subscribe(response => {
            this.bootstrapper.bootstrap(response.data);

            if ( this.activeRoute.snapshot.queryParams['action'] === 'add_site') {
              const projectName: string = this.activeRoute.snapshot.queryParams['project_name'];

              this.router.navigate(['dashboard/projects/new'], { queryParams: { project_name: projectName } });
            } else if ( this.activeRoute.snapshot.queryParams['action'] === 'edit_site') {
              const projectName: string = this.activeRoute.snapshot.queryParams['project_name'];

              this.router.navigate(
                ['builder', this.activeRoute.snapshot.queryParams['project_id']],
                { queryParams: { project_name: projectName } }
              );
            } else {
              //TODO: Move this into auth service, so other components can re-use
              this.router.navigate([this.auth.getRedirectUri()]).then(navigated => {
                this.isLoading = false;

                if (!navigated) {
                  this.router.navigate([this.auth.getRedirectUri()]);
                }
              })
            }
        }, response => {
            this.errors = response['messages'];
            this.isLoading = false;
        });
    }

    private hydrateModel() {
        if (this.settings.get('vebto.site.demo')) {
            this.model.email = 'admin@admin.com';
            this.model.password = 'admin';
        }
    }
}
